### ISSUE TITLE
Segmentation fault with new apache httpd 2.4.54

### ISSUE THREAD

**Issue by uhliarik** [GrahamDumpleton/mod_wsgi#788]
Hi Graham,

with the newest httpd 2.4.54 and possibly because of this commit (https://github.com/apache/httpd/commit/ce259c4061905bf834f9af51c92456cfe8335ddc), we are hitting segmentation faults.

bt looks following:

```
#0  0x00007f03dad87d7d in __strlen_avx2 () from /lib64/libc.so.6
#1  0x00007f03da01dacf in wsgi_header_filter (f=0x7f03cc0061b8, b=0x7f03cc008aa8) at src/server/mod_wsgi.c:12550
#2  0x00005604000bb586 in ap_http_filter (f=<optimized out>, b=<optimized out>, mode=AP_MODE_READBYTES, block=APR_BLOCK_READ, readbytes=8192) at modules/http/http_filters.c:644
#3  0x00007f03da004fd4 in Input_read_from_input (self=self@entry=0x7f03c303f280, 
    buffer=buffer@entry=0x7f03cd0ee0d0 "\224n\210\001\246\306\344\303\351c\207\237\226\304\377\327\363\200eϧ\003έ/\310\r\372?\346n\331T\004\020\034L-\206t\202!1\304\350\325a\t\260[\214\004>\300\247b\312\b[E\266\252\037\301/\361\244\202\235\276\273|~", bufsiz=<optimized out>) at src/server/mod_wsgi.c:1130
#4  0x00007f03da00541a in Input_read (self=0x7f03c303f280, args=<optimized out>) at src/server/mod_wsgi.c:1366
```

We are trying to upload a big amount of data (4GB) through python application running on mod_wsgi, but in frame 2 we are hitting condition, that read_data are greater than limit. 

http_filters.c:
```
                /* We have a limit in effect. */
                if (ctx->limit) {
                    /* FIXME: Note that we might get slightly confused on
                     * chunked inputs as we'd need to compensate for the chunk
                     * lengths which may not really count.  This seems to be up
                     * for interpretation.
                     */
                    ctx->limit_used += totalread;
                    if (ctx->limit < ctx->limit_used) {
                        ap_log_rerror(APLOG_MARK, APLOG_INFO, 0, f->r,
                                      APLOGNO(01591) "Read content length of "
                                      "%" APR_OFF_T_FMT " is larger than the "
                                      "configured limit of %" APR_OFF_T_FMT,
                                      ctx->limit_used, ctx->limit);
                        return bail_out_on_error(ctx, f,
                                                 HTTP_REQUEST_ENTITY_TOO_LARGE);
                    }
                }
```

If you look at the beginning of the ap_http_filter function, you can see how ctx->limit is set:
```
    if (!ctx) {
        const char *tenc, *lenp;
        f->ctx = ctx = apr_pcalloc(f->r->pool, sizeof(*ctx));
        ctx->state = BODY_NONE;
        ctx->bb = apr_brigade_create(f->r->pool, f->c->bucket_alloc);

        /* LimitRequestBody does not apply to proxied responses.
         * Consider implementing this check in its own filter.
         * Would adding a directive to limit the size of proxied
         * responses be useful?
         */
        if (!f->r->proxyreq) {
            ctx->limit = ap_get_limit_req_body(f->r);
        }
        else {
            ctx->limit = 0;
        }
```

Since before, f->r->per_fir_config->limit_req_body has been always set to -1 (uninitialized), ap_get_limit_req_body function automatically set ctx->limit to 0 (unlimited), but now after the change in httpd apache it changes ctx->limit to 1GiB. 

Anyway although it sets ctx->limit to 1 GiB, I don't understand why after returning the error, mod_wsgi is segfaulting on this line:

mod_wsgi.c (wsgi_header_filter):
```
    vec1[2].iov_len  = strlen(r->status_line);    // <----  r->status_line is null.
```

I'm trying to find out what is wrong but I'm not so familiar with mod_wsgi code. Your help is welcome. If you need any additional info, please let me know.

**Comment by GrahamDumpleton** [comment]
That code line should only be executed in the mod_wsgi daemon process. In the case of `LimitRequestBody` being exceeded, the request shouldn't even be getting proxied across to the daemon process as it should be blocked in the Apache child process before sending. So would be strange if that is what is occurring.

For now if you set `LimitRequestBody` directive back to 0 explicitly, does it at least work as before?

**Comment by GrahamDumpleton** [comment]
The code executed in Apache child process which aborts handling (without proxying) is:

```
    /*
     * Check to see if the request content is too large if the
     * Content-Length header is defined then end the request here. We do
     * this as otherwise it will not be done until first time input data
     * is read in by the application. Problem is that underlying HTTP
     * output filter will also generate a 413 response and the error
     * raised from the application will be appended to that. The call to
     * ap_discard_request_body() is hopefully enough to trigger sending
     * of the 413 response by the HTTP filter.
     */

    lenp = apr_table_get(r->headers_in, "Content-Length");

    if (lenp) {
        char *endstr;
        apr_off_t length;

        if (wsgi_strtoff(&length, lenp, &endstr, 10)
            || *endstr || length < 0) {

            wsgi_log_script_error(r, apr_psprintf(r->pool,
                    "Invalid Content-Length header value of '%s' was "
                    "supplied.", lenp), r->filename);

            return HTTP_BAD_REQUEST;
        }

        limit = ap_get_limit_req_body(r);

        if (limit && limit < length) {
            ap_discard_request_body(r);
            return OK;
        }
    }
```

**Comment by uhliarik** [comment]
Thanks for your answer Graham. 

That the problem, that I can not set LimitRequestBody back to the 0. I tried to set it, but it seems to me that there are some "fake requests" sent by mod_wsgi, which have uninitialized structures.

```
(gdb) p *(core_dir_config *)((void **)(f->r->per_dir_config))[0]
$14 = {d = 0x0, d_components = 0, opts = 20 '\024', opts_add = 0 '\000', opts_remove = 0 '\000', override = 32, override_opts = 255 '\377', response_code_strings = 0x0, 
  hostname_lookups = 3, content_md5 = 2, use_canonical_name = 3, d_is_fnmatch = 0, add_default_charset = 1, add_default_charset_name = 0x5604014572f0 "UTF-8", limit_cpu = 0x0, 
  limit_mem = 0x0, limit_nproc = 0x0, limit_req_body = -1, limit_xml_body = -1, server_signature = srv_sig_unset, sec_file = 0x5604014579d8, sec_if = 0x0, r = 0x0, mime_type = 0x0, 
  handler = 0x0, output_filters = 0x0, input_filters = 0x0, 	accept_path_info = 3, etag_bits = 0, etag_add = 0, etag_remove = 0, enable_mmap = 2, enable_sendfile = 1, 
  use_canonical_phys_port = 2, allow_encoded_slashes = 0, decode_encoded_slashes = 0, condition_ifelse = 0, condition = 0x0, log = 0x560401397db0, override_list = 0x0, max_ranges = -1, 
  max_overlaps = -1, max_reversals = -1, refs = 0x0, response_code_exprs = 0x0, cgi_pass_auth = 2, qualify_redirect_url = 2, expr_handler = 0x0, cgi_var_rules = 0x0, read_buf_size = 0}
```

Therefore here https://github.com/apache/httpd/blob/2.4.x/modules/http/http_filters.c#L371 the limit is set to default (in case of new httpd 2.4.54 it is 1GiB instead of unlimited). As you can see in https://github.com/apache/httpd/blob/2.4.x/server/core.c#L1248 , in this function it reads f->r->per_dir_config which has always set  f->r->per_dir_config->d to NULL in case of mod_wsgi "fake "request"

**Comment by GrahamDumpleton** [comment]
When you say fake requests, are you talking about how mod_wsgi emulates creation of an Apache request in the mod_wsgi daemon process? This is only done so same Apache request handling mechanism can be used in the mod_wsgi daemon process as is used in Apache child processes. As mentioned before though, if the request body limit is being exceeded the request shouldn't even get proxied to the daemon process.

Maybe the issue is that they introduced a new not set value for the limit (-1), which didn't exist before, and so this check is wrong:

```
        if (limit && limit < length) {
            ap_discard_request_body(r);
            return OK;
        }
```

In other words, should be:

```
        if (limit > 0 && limit < length) {
            ap_discard_request_body(r);
            return OK;
        }
```

Can you provide the complete stack trace from the core dump so can verify which process it is occurring in?

**Comment by GrahamDumpleton** [comment]
Problem is am not sure what should be doing about not set case. I would need to know exactly which Apache version they made change and have different behaviour, with it using the compiled in default they have elsewhere if is new enough Apache. So that change in itself would not be enough.

**Comment by GrahamDumpleton** [comment]
Actually that change should matter since they use:

```
AP_DECLARE(apr_off_t) ap_get_limit_req_body(const request_rec *r)
{
    core_dir_config *d =
      (core_dir_config *)ap_get_core_module_config(r->per_dir_config);

    if (d->limit_req_body == AP_LIMIT_REQ_BODY_UNSET) {
        return AP_DEFAULT_LIMIT_REQ_BODY;
    }

    return d->limit_req_body;
}
```

So they return the new limit if unset automatically. So wouldn't see -1. So still can't see how request would get proxied and that other code executed in header filter.

**Comment by GrahamDumpleton** [comment]
Are you by chance using compressed request content and non standard WSGI behaviour of reading until end of stream instead of adhering to length. I can see maybe this could throw things out if using daemon mode.

**Comment by GrahamDumpleton** [comment]
Since is not obvious how getting into the situation, only thing can think to do is if `r->status_line` is `NULL`, is to set it to:

```
ap_get_status_line(r->status)
```

This will at least give it a value.

**Comment by GrahamDumpleton** [comment]
If you are able to, can you try the "develop" branch of https://github.com/GrahamDumpleton/mod_wsgi

**Comment by GrahamDumpleton** [comment]
I think am starting to understand. In daemon mode process it relied on limit effectively being unlimited as it didn't inherit configuration value even if set in Apache config. If however that is now been overridden by filter when unset, it will affect the daemon process. So when the emulate request setup is done, have to ensure limit value is set 0 and not -1. Problem is finding where that is set.

**Comment by notroj** [comment]
FYI @GrahamDumpleton we added `ap_create_request()` in httpd 2.4.49 specifically to address the `mod_wsgi` case of trying to create a `request_rec`, I don't know if it would avoid this particular issue, but it should at least avoid a repeat of the issue we had with `CVE-2013-5704`.

**Comment by GrahamDumpleton** [comment]
Don't remind me of CVE-2013-5704. Ran across that in the code as was looking through it and groaned as a I remembered and was worried I would when looking back not even remember what some of these hacks were. :-)

Right now am thinking of cheating. Since in daemon process and can't affect main config in Apache child processes, am going to do something like:

```
core_dir_config *d = (core_dir_config *)ap_get_core_module_config(r->per_dir_config);
d-> limit_req_body = 0;
```

I'll look at `ap_create_request()` and see if there is anything can learn from that.

**Comment by GrahamDumpleton** [comment]
@uhliarik Have made another speculative change.

```
    /*
     * Try and ensure that request body limit in daemon mode process
     * is unlimited as Apache 2.4.54 changed rules for limit and if
     * unset is now overridden by HTTP filters to be 1GiB rather than
     * unlimited. Not sure if this is required or not but could be
     * since we populate configuration from base server config only.
     */

    d = (core_dir_config *)ap_get_core_module_config(r->per_dir_config);

    d->limit_req_body = 0;
```

So what may be happening is that in general in mod_wsgi daemon mode the Apache configuration is ignored as everything that depended on it should have been handled in Apache child process where proxying is initiated.

To satisfy emulating Apache request handling, the base server config was used when creating the fake request. If the HTTP filter which is still used in daemon mode consults the config for request body limit now, it will only look at base server config. So if you have overridden request body limit in virtual host, or a directory/location configuration to set it to 0, that wouldn't be seen and used.

So above change is taking a rather brute force approach which hope may work, which is to override the `limit_req_body` and set it to 0 on every request. If this worked, could probably work out place to only override it the once for process after fork, but will do for now.

**Comment by uhliarik** [comment]
I'm adding requested backtrace:

```
(gdb) bt
#0  0x00007f03dad87d7d in __strlen_avx2 () from /lib64/libc.so.6
#1  0x00007f03da01dacf in wsgi_header_filter (f=0x7f03cc0061b8, b=0x7f03cc008aa8) at src/server/mod_wsgi.c:12550
#2  0x00005604000bb586 in ap_http_filter (f=<optimized out>, b=<optimized out>, mode=AP_MODE_READBYTES, block=APR_BLOCK_READ, readbytes=8192) at modules/http/http_filters.c:644
#3  0x00007f03da004fd4 in Input_read_from_input (self=self@entry=0x7f03c303f280, 
    buffer=buffer@entry=0x7f03cd0ee0d0 "\224n\210\001\246\306\344\303\351c\207\237\226\304\377\327\363\200eϧ\003έ/\310\r\372?\346n\331T\004\020\034L-\206t\202!1\304\350\325a\t\260[\214\004>\300\247b\312\b[E\266\252\037\301/\361\244\202\235\276\273|~", bufsiz=<optimized out>) at src/server/mod_wsgi.c:1130
#4  0x00007f03da00541a in Input_read (self=0x7f03c303f280, args=<optimized out>) at src/server/mod_wsgi.c:1366
#5  0x00007f03d9dc10e8 in cfunction_call () from /lib64/libpython3.10.so.1.0
#6  0x00007f03d9dbd2b4 in _PyObject_MakeTpCall () from /lib64/libpython3.10.so.1.0
#7  0x00007f03d9dba07e in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#8  0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#9  0x00007f03d9dc9188 in method_vectorcall () from /lib64/libpython3.10.so.1.0
#10 0x00007f03d9db41c9 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#11 0x00007f03d9dcd4cd in gen_send_ex2 () from /lib64/libpython3.10.so.1.0
#12 0x00007f03d9dcd398 in gen_iternext () from /lib64/libpython3.10.so.1.0
#13 0x00007f03d9e1dad9 in chain_next () from /lib64/libpython3.10.so.1.0
#14 0x00007f03d9db4938 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#15 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#16 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#17 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#18 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#19 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#20 0x00007f03d9db41c9 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#21 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#22 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#23 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#24 0x00007f03d9db9456 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#25 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#26 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#27 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#28 0x00007f03d9db9456 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#29 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#30 0x00007f03d9dae966 in _PyObject_VectorcallTstate.lto_priv.5 () from /lib64/libpython3.10.so.1.0
#31 0x00007f03d9e30697 in object_vacall () from /lib64/libpython3.10.so.1.0
#32 0x00007f03d9e30571 in PyObject_CallFunctionObjArgs () from /lib64/libpython3.10.so.1.0
#33 0x00007f03d9dc0b3d in _PyObject_GenericGetAttrWithDict () from /lib64/libpython3.10.so.1.0
#34 0x00007f03d9dc0108 in PyObject_GetAttr () from /lib64/libpython3.10.so.1.0
#35 0x00007f03d9dc1720 in cfunction_vectorcall_FASTCALL () from /lib64/libpython3.10.so.1.0
#36 0x00007f03d9dae966 in _PyObject_VectorcallTstate.lto_priv.5 () from /lib64/libpython3.10.so.1.0
#37 0x00007f03d9e63060 in partial_vectorcall () from /lib64/libpython3.10.so.1.0
#38 0x00007f03d9dbfbdd in _PyObject_VectorcallTstate.lto_priv.6 () from /lib64/libpython3.10.so.1.0
#39 0x00007f03d9e66e39 in call_attribute () from /lib64/libpython3.10.so.1.0
#40 0x00007f03d9e070da in slot_tp_getattr_hook () from /lib64/libpython3.10.so.1.0
#41 0x00007f03d9dc011a in PyObject_GetAttr () from /lib64/libpython3.10.so.1.0
#42 0x00007f03d9db91df in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#43 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#44 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
--Type <RET> for more, q to quit, c to continue without paging--f 1
#45 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#46 0x00007f03d9dc9188 in method_vectorcall () from /lib64/libpython3.10.so.1.0
#47 0x00007f03d9dc9760 in PyVectorcall_Call () from /lib64/libpython3.10.so.1.0
#48 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#49 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#50 0x00007f03d9dc9188 in method_vectorcall () from /lib64/libpython3.10.so.1.0
#51 0x00007f03d9dc9760 in PyVectorcall_Call () from /lib64/libpython3.10.so.1.0
#52 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#53 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#54 0x00007f03d9dbcaf1 in _PyObject_FastCallDictTstate () from /lib64/libpython3.10.so.1.0
#55 0x00007f03d9dc7813 in slot_tp_init () from /lib64/libpython3.10.so.1.0
#56 0x00007f03d9dbd533 in type_call () from /lib64/libpython3.10.so.1.0
#57 0x00007f03d9e07a8f in wrapperdescr_call () from /lib64/libpython3.10.so.1.0
#58 0x00007f03d9dc9628 in _PyObject_Call () from /lib64/libpython3.10.so.1.0
#59 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#60 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#61 0x00007f03d9dbcaf1 in _PyObject_FastCallDictTstate () from /lib64/libpython3.10.so.1.0
#62 0x00007f03d9dc7a1c in _PyObject_Call_Prepend () from /lib64/libpython3.10.so.1.0
#63 0x00007f03d9e6af72 in slot_tp_call () from /lib64/libpython3.10.so.1.0
#64 0x00007f03d9dbd2b4 in _PyObject_MakeTpCall () from /lib64/libpython3.10.so.1.0
#65 0x00007f03d9db96ce in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#66 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#67 0x00007f03d9db41c9 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#68 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#69 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#70 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#71 0x00007f03d9dc9760 in PyVectorcall_Call () from /lib64/libpython3.10.so.1.0
#72 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#73 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#74 0x00007f03d9dc9760 in PyVectorcall_Call () from /lib64/libpython3.10.so.1.0
#75 0x00007f03d9db6e98 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#76 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#77 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#78 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#79 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#80 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#81 0x00007f03d9db4640 in _PyEval_EvalFrameDefault () from /lib64/libpython3.10.so.1.0
#82 0x00007f03d9db3000 in _PyEval_Vector () from /lib64/libpython3.10.so.1.0
#83 0x00007f03d9dbcaf1 in _PyObject_FastCallDictTstate () from /lib64/libpython3.10.so.1.0
#84 0x00007f03d9dc7a1c in _PyObject_Call_Prepend () from /lib64/libpython3.10.so.1.0
#85 0x00007f03d9e6af72 in slot_tp_call () from /lib64/libpython3.10.so.1.0
#86 0x00007f03d9dc9628 in _PyObject_Call () from /lib64/libpython3.10.so.1.0
#87 0x00007f03da00ddc7 in Adapter_run (object=0x7f03d98d2f20, self=<optimized out>) at src/server/mod_wsgi.c:3182
#88 wsgi_execute_script (r=<optimized out>) at src/server/mod_wsgi.c:4215
#89 0x00007f03da0180f6 in wsgi_hook_daemon_handler (c=<optimized out>) at src/server/mod_wsgi.c:13233
#90 wsgi_process_socket (daemon=0x56040144aec0, bucket_alloc=<optimized out>, sock=<optimized out>, p=<optimized out>) at src/server/mod_wsgi.c:8732
#91 wsgi_daemon_worker (thread=<optimized out>, p=<optimized out>) at src/server/mod_wsgi.c:9060
#92 wsgi_daemon_thread (thd=<optimized out>, data=<optimized out>) at src/server/mod_wsgi.c:9150
#93 0x00007f03dad698ca in start_thread () from /lib64/libc.so.6
#94 0x00007f03dad09500 in clone3 () from /lib64/libc.so.6
```

**Comment by uhliarik** [comment]
> is worked, could probably work out place to only override it the once for process after fork, but will do for now.

I think this looks reasonable. I was looking for some place in the code where I could force this limit_req_body to 0 (unlimited). Thanks for finding it. I will try to rebuild mod_wsgi today with this change and I will post here if it worked.

Anyway this may only fix the issue with ignoring limit_req_body but I'm not sure if the mod_wsgi process should segfault. But we can try to solve it later.

**Comment by uhliarik** [comment]
> Are you able to try with develop version of code?

OK, I will try to create build with develop version.

**Comment by GrahamDumpleton** [comment]
Is getting late here now, but now that I believe I understand exactly how it is coming about, if I get a chance tomorrow I should be able to replicate the issue myself to confirm the fix. If current change doesn't help, will have to work out a different way.

**Comment by uhliarik** [comment]
> Is getting late here now, but now that I believe I understand exactly how it is coming about, if I get a chance tomorrow I should be able to replicate the issue myself to confirm the fix. If current change doesn't help, will have to work out a different way.

OK no worries, I saw your changes in devel branch. I will try to create a build from this branch and test it. I saw you also addressed r->status_line == NULL issue.

**Comment by uhliarik** [comment]
I tried to build mod_wsgi with both changes separately first and the good news is that if mod_wsgi contains only the change with r->status_line initialization in case r->status_line is NULL then mod_wsgi won't segfault. Anyway it stops wrongly uploading after 1GiB although no limit is set. 

But if I also include second change where limit_req_body is initialized to 0 (unlimited), then mod_wsgi behaves correctly.

Semantically the change makes sense to me as well.

Thanks Graham.

**Comment by uhliarik** [comment]
I would like to add, that this issue occurs only if https(mod_ssl) is used.

**Comment by GrahamDumpleton** [comment]
Technically I can't see how it wouldn't also affect HTTP only. So you may just have been lucky. :-)

**Comment by GrahamDumpleton** [comment]
Fixes release in version 4.9.4.
